### PR TITLE
Migrate vm across clusters

### LINKED ISSUE #3707: When only one host is available and it is put in maintenance, all VMs on it are Stopped.
When host is put in maintenance, and it is the only host available, all VMs on it are shutdown. This seems intentional code implementation and needs change.

##### ISSUE TYPE

 * Improvement Request

##### COMPONENT NAME

~~~

~~~

##### CLOUDSTACK VERSION

~~~

~~~

##### STEPS TO REPRODUCE

~~~
Go to host on a cluster with single host, or all but one host put into maintenance, errorinmaintenance state. Now put the last host in maintenance.
~~~

##### EXPECTED RESULTS

~~~
VMs are not stopped and proper message is displayed. Host should be put to ErrorInMaintenance State.
~~~

##### ACTUAL RESULTS

All VMs are shutdown and host is put to maintenance.

~~~

~~~

### PR DESCRIPTION
**Pull request by ustcweizhou**
### Description

This PR adds a global setting migrate.vm.across.clusters to indicate whether vm can be live migrated to other clusters.

If vm is running on the last host in a cluster and vm cannot be migrated to other clusters, put host to maintenance will fail.

This is based on pr #4378 

This fixes #3707 #3720 

Some rules
(1) for vmware, across-cluster migration of vms with cluster-scoped pools is supported
(2) for other hypervisors except vmware, vm can be live migrated to other clusters (with same hypervisor type), if all volumes are in zone-wide storage pool.
(3) migration of systemvms (CPVM, SSVM) is only possible across clusters in same pod to avoid potential network issues.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

### CODE REVIEW

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -5787,6 +5789,50 @@ public VirtualMachine migrateVirtualMachine(Long vmId, Host destinationHost) thr
...
+            collectVmDiskStatistics(uservm);
+            collectVmNetworkStatistics(uservm);
+        }
+        _itMgr.migrate(vm.getUuid(), srcHostId, dest);
+        VMInstanceVO vmInstance = _vmInstanceDao.findById(vmId);
+        if (vmInstance.getType().equals(VirtualMachine.Type.User)) {
+            return _vmDao.findById(vmId);
+        } else {
+            return vmInstance;
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: I know i sound like a broken record , but I see four methods here. Can you extract these pieces of code please?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -5812,7 +5814,7 @@ public VirtualMachine migrateVirtualMachine(Long vmId, Host destinationHost) thr
 
         // If no suitable destination found then throw exception
         if (dest == null) {
-            throw new RuntimeException("Unable to find suitable destination to migrate VM " + vm.getInstanceName());
+            throw new CloudRuntimeException("Unable to find suitable destination to migrate VM " + vm.getInstanceName());
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:5815]: 👍

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java [hunk 3]
```diff
@@ -1262,7 +1263,29 @@ private boolean doMaintain(final long hostId) {
...
+                        if (_vmMgr.checkIfVmHasClusterWideVolumes(vm.getId())) {
+                            s_logger.warn("Unable to migrate vm " + vm.getInstanceName() + " as it has volumes on cluster-wide pool");
+                            return false;
+                        }
+                    }
+                } else {
+                    s_logger.warn("Not migrating VM across cluster since " + MIGRATE_VM_ACROSS_CLUSTERS.key() + " is false");
+                    return false;
+                }
+            }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/resource/ResourceManagerImpl.java]: sugestion:
`if (! clusterWideMigrationSupported(...)) return false;`
and extract this code to
`boolean clusterWideMigrationSupported(...)`

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 4]
```diff
@@ -3268,6 +3226,22 @@ public boolean checkIfVmHasClusterWideVolumes(Long vmId) {
 
     }
 
+    private DataCenterDeployment getMigrationDeployment(final Long vmId, final Host host, final Long poolId, final ExcludeList excludes) {
+        if (MIGRATE_VM_ACROSS_CLUSTERS.value() && !checkIfVmHasClusterWideVolumes(vmId)) {
+            s_logger.info("Searching for hosts in the zone for vm migration");
+            List<Long> clustersToExcluded = _clusterDao.listAllClusters(host.getDataCenterId());
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:3233]: sugest you name this `clustersToExclude` without the d on the end

**Review by DaanHoogland — APPROVED** [review]
thanks for taking this forward @weizhouapache . I have my usual style complaints but looks good otherwise. I asume you are already running this in your env, I wonder if the functional reqs on this in thers envs are met but that is for them to asure. (no tests done)

**Comment by shwstppr** [comment]
@blueorangutan package

@weizhouapache is this ready?

**Comment by shwstppr** [comment]
@ustcweizhou I guess this will need changes in new UI now
I've not tested yet but cross-cluster migration could result in failure when there are only cluster-scoped primary stores so do we handle that? In case of VMware this could fail even with vMotion enabled with current master something that has been changed in #4385

**Comment by shwstppr** [comment]
Based on branch name marked for 4.15.1.0

**Comment by weizhouapache** [comment]
> @ustcweizhou I guess this will need changes in new UI now
> I've not tested yet but cross-cluster migration could result in failure when there are only cluster-scoped primary stores so do we handle that? In case of VMware this could fail even with vMotion enabled with current master something that has been changed in #4385

@shwstppr it is addressed in this pr.  if vm has volumes on cluster-wide storage, migrate will fail.
migraiton is possible only if
(1) all volumes are on zone-wide storage
(2) source cluster and dest cluster have same hypervisor type.

**Comment by weizhouapache** [comment]
@DaanHoogland 
this is an improvement so I changed the destination to master.

made some changes as your comments.

**Comment by shwstppr** [comment]
@weizhouapache sorry but this needs fixing conflicts again.
Also, is the use-case here is allow cross-cluster migration when putting the last host in maintenance. As we added some improvements for cross-cluster cross-pod VM migration with #4385 but they were largely for VMware.
Since VMware would allow cross-cluster migration even with cluster scoped pools with vmotion do we need to mention the new global setting is not for VMware

**Comment by rhtyd** [comment]
Re-ping @shwstppr for review
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
> sorry but this needs fixing conflicts again.
> Also, is the use-case here is allow cross-cluster migration when putting the last host in maintenance. As we added some improvements for cross-cluster cross-pod VM migration with #4385 but they were largely for VMware.
> Since VMware would allow cross-cluster migration even with cluster scoped pools with vmotion do we need to mention the new global setting is not for VMware

@shwstppr 
I had a quick view on #4385, it looks that it migrates vms with a storageid.
this pr should not break the feature as this changes the behavior when hostid is passed or storageid is not passed.

**Comment by rhtyd** [comment]
Ping @weizhouapache is this valid for 4.15, in terms of severity should this be considered for 4.15.1?

**Comment by weizhouapache** [comment]
> Ping @weizhouapache is this valid for 4.15, in terms of severity should this be considered for 4.15.1?

@rhtyd it is more like a new feature, so targeted to master/4.16, not 4.15

**Comment by rhtyd** [comment]
Thanks @weizhouapache - this was regards to https://github.com/apache/cloudstack/issues/3720 which is logged as a bug which this PR claims to fix.

**Comment by rhtyd** [comment]
ping @shwstppr is this relevant, I think there was a PR from you as well on similar (cross-cluster migration)
@ustcweizhou can you fix the conflicts?

**Comment by sureshanaparti** [comment]
Hi @ustcweizhou can you please resolve the conflicts.

**Comment by weizhouapache** [comment]
> Hi @ustcweizhou can you please resolve the conflicts.

@sureshanaparti 
This is based on pr #4378
I will rebase when #4378 is merged.

**Comment by soreana** [comment]
@ustcweizhou should we take care of this?

**Comment by weizhouapache** [comment]
> @ustcweizhou should we take care of changes for this pr?

@soreana #4378 is not merged yet.

**Comment by soreana** [comment]
@weizhouapache I see. Let me know if you need an extra hands :)

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@weizhouapache can you please fix the conflicts? Is it ready for review?

**Comment by soreana** [comment]
@nvazquez I've fixed the conflict. It is ready for review.

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/deploy/FirstFitPlanner.java [hunk 5]
```diff
@@ -441,7 +441,22 @@ protected void removeClustersCrossingThreshold(List<Long> clusterListForVmAlloca
...
+            List<Long> clustersWithDisabledPods = clusterDao.listClustersWithDisabledPods(zoneId);
+            disabledClusters.addAll(clustersWithDisabledPods);
+        }
+        return disabledClusters;
+    }
+
+    private List<Long> listDisabledPods(long zoneId) {
+        List<Long> disabledPods = podDao.listDisabledPods(zoneId);
+        return disabledPods;
+    }
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/deploy/FirstFitPlanner.java:457]: @weizhouapache these methods are not used?

**Review by shwstppr — CHANGES_REQUESTED** [review]
@weizhouapache can you please test the PR. In my testing migration didn't work with the following setup:
- KVM
- 2x clusters (1x host each, 1x cluster scope primary store each)
- 1x zone-wide store
- `migrate.vm.across.clusters` set to true
- VM that didn't migrate had its volume in zone wide pool
Logs:
```
2021-09-03 05:53:58,467 WARN  [c.c.r.ResourceManagerImpl] (API-Job-Executor-4:ctx-a66a8e8f job-71 ctx-6acba5aa) (logid:53d20614) Unable to find a host for vm migration in cluster: 2
2021-09-03 05:53:58,471 ERROR [c.c.r.ResourceManagerImpl] (API-Job-Executor-4:ctx-a66a8e8f job-71 ctx-6acba5aa) (logid:53d20614) Maintenance: No hosts available for migrations. Scheduling shutdown for VM 681b9a30-fe81-4104-b9e0-3036165c5065 instead of migration.
```

Also, I feel:

- migration should be restricted to within pod at least for systemvm as there could be an issue with the private IP in that case.
- for VMware, maybe we can skip volumes in cluster pool check as it can allow migration across clusters with cluster-wide pools as well

**Comment by weizhouapache** [comment]
@shwstppr thanks for review and testing !

as per your comments, I have changed this pr by 5 commits.

fix #4534: an error in 'git merge'
fix #4534: remove useless methods in FirstFitPlanner.java
fix #4534: vms are stopped in host maintenance
fix #4534: across-cluster migration of vms with cluster-scoped pools is supported by vmware vmotion
fix #4534: migrate systemvms is only possible across clusters in same pod to avoid potential network errors.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@shwstppr great, thanks for testing !

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java [hunk 6]
```diff
@@ -1345,6 +1353,37 @@ private boolean doMaintain(final long hostId) {
...
+                return false;
+            }
+            s_logger.info("Found hosts in the zone for vm migration: " + hosts);
+            // Don't migrate vm if it has volumes on cluster-wide pool
+            for (final VMInstanceVO vm : vms) {
+                if (! HypervisorType.VMware.equals(host.getHypervisorType()) && _vmMgr.checkIfVmHasClusterWideVolumes(vm.getId())) {
+                    s_logger.warn(String.format("VM %s cannot be migrated across cluster as it has volumes on cluster-wide pool", vm));
+                    return false;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java]: @shwstppr 
moved the hypervisor type check.

I have considered the order of checks. in my opinion, the checks of vm volumes have more db queries than listAllUpAndEnabledHosts so I put the check prior to volume pool check.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by nvazquez — APPROVED** [review]
LGTM

**Pull request merged**